core/types/calls.cc

```diff
@@ -1916,6 +1916,45 @@ class Array_flatten : public IntrinsicMethod {
     }
 } Array_flatten;
 
+class Array_product : public IntrinsicMethod {
+public:
+    void apply(Context ctx, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        // Unwrap the array one time to get the element type (we'll rewrap it down at the bottom)
+        TypePtr element;
+        if (auto *ap = cast_type<AppliedType>(thisType)) {
+            ENFORCE(ap->klass == Symbols::Array() || ap->klass.data(ctx)->derivesFrom(ctx, Symbols::Array()));
+            ENFORCE(!ap->targs.empty());
+            element = ap->targs.front();
+        } else if (auto *tuple = cast_type<TupleType>(thisType)) {
+            element = tuple->elementType();
+        } else {
+            ENFORCE(false, "Array#product on unexpected type: {}", args.selfType->show(ctx));
+        }
+
+        vector<TypePtr> unwrappedElems;
+        unwrappedElems.reserve(args.args.size() + 1);
+        unwrappedElems.emplace_back(element);
+
+        for (auto arg : args.args) {
+            auto argTyp = arg->type;
+            TypePtr el;
+            if (auto *ap = cast_type<AppliedType>(argTyp.get())) {
+                ENFORCE(ap->klass == Symbols::Array() || ap->klass.data(ctx)->derivesFrom(ctx, Symbols::Array()));
+                ENFORCE(!ap->targs.empty());
+                el = ap->targs.front();
+            } else if (auto *tuple = cast_type<TupleType>(argTyp.get())) {
+                el = tuple->elementType();
+            } else {
+                ENFORCE(false, "Array#product called with unexpected type: {}", el->show(ctx));
+            }
+
+            unwrappedElems.emplace_back(el);
+        }
+
+        res.returnType = Types::arrayOf(ctx, TupleType::build(ctx, unwrappedElems));
```

Comment:
I'm not really sure what `Array#product` does. Can you explain why this is the right type?

Reply:
Answering my own question: it looks like the number of args always corresponds to the length of the tuple, and other than that, the nesting depth of arrays in the result is always 2: the array of all combinations, containing tuples within.

Reply:
`Array#product` returns the cartesian products of the `this` array as well as all arrays passed in.

For arrays `A` and `B`, the result of `A.product(B)` is  `[(x, y) : x ∈ A, y ∈ B]`